docs/Compiler-Hardening-Guides/Compiler-Options-Hardening-Guide-for-C-and-C++.md

```diff
@@ -394,6 +396,52 @@ Some tools, such as `autoconf`, automatically determine what the compiler suppor
+
+For more about Trojan Source see [trojansource.com](https://trojansource.codes/), which includes a copy of the paper "Trojan Source: Invisible Vulnerabilities" by Nicholas Boucher and Ross Anderson [^Boucher2021]. More context can be found [this Hacker News discussion](https://news.ycombinator.com/item?id=29062982) and [Wikipedia](https://en.wikipedia.org/wiki/Trojan_Source).  See, in particular, [CVE-2021-42574](https://nvd.nist.gov/vuln/detail/CVE-2021-42574) and [CVE-2021-42694](https://nvd.nist.gov/vuln/detail/CVE-2021-42684).
+
+GCC has a `-Wbidi-chars` option to help counter Trojan Source attacks [^gcc-Wbidi-chars].  By default its value is `-Wbidi-char=unpaired`, which warns about improperly terminated bidi contexts (this should never happen in source code). However, this default is somewhat permissive.
+
+In many cases using `-Wbidi-chars=any` is a stronger defense. This option forbids *any* use of bidirectional control characters, completely eliminating the Trojan Source attack. This setting is appropriate when such characters are *not* expected in the source code, and their only use would be as part of an attack on reviewers. This implements a "least privilege" mechanism for source code processing - if bidi characters are not expected, there's no need to permit them.
+
+Note that this option does *not* interfere with creating internationalized programs. Current best practice is to put human-readable text strings in separate file(s), not in source code, and then use an internationalization (i18n) framework like `gettext` to retrieve the correct text for the user's locale.
+
+<!-- Implemented in: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=103026 -->
+
+[^Boucher2021]: [Boucher, Nicholas and Ross Anderson, 2021, "Trojan Source: Invisible Vulnerabilities"](https://arxiv.org/abs/2111.00169)
+
+[^gcc-Wbidi-chars]: GCC team, [Using the GNU Compiler Collection (GCC): Warning Options: `-Wbidi-chars`](https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wbidi-chars_003d),
+
+[^Wheeler2020]: [Wheeler, David A., April 2020, *Initial Analysis of Underhanded Source Code*](https://www.ida.org/-/media/feature/publications/i/in/initial-analysis-of-underhanded-source-code/d-13166.ashx).
+
+#### Performance implications
+
+None.
+
+#### When not to use?
+
+Do *not* use `-Wbidi-chars=any` in cases where some of the source code *is* expected to include bidirectional control characters. This is typically the case where some of the source code text is in Arabic or Hebrew. In such cases, use `-Wbidi-chars=unpaired` (the default) instead on GCC. This alternative still provides some defenses.
```

Comment:
Suggest rewording to 

> Arabic, Hebrew, or other right-to-left script.

Reply:
I reworded further. It turns out that Persian and Urdu are right-to-left. We'll mention both, in both lists of examples. That'll at least make it clear there are *multiple* writing systems relevant to this (and thus why "always forbidding" isn't the right approach).